components/tikv_util/src/logger/file_log.rs

```diff
@@ -1,196 +1,400 @@
-        let file = open_log_file(&file_path)?;
-        let file_attr = fs::metadata(&file_path)?;
-        let file_modified_time = file_attr.modified().unwrap().into();
-        let next_rotation_time = compute_rotation_time(&file_modified_time, rotation_timespan);
-        Ok(Self {
-            next_rotation_time,
-            file_path,
-            rotation_timespan,
+/// Builder for `RotatingFileLogger`.
+pub struct RotatingFileLoggerBuilder<N>
+where
+    N: 'static + Send + Fn(&Path) -> io::Result<PathBuf>,
+{
+    rotators: Vec<Box<dyn Rotator>>,
+    path: PathBuf,
+    rename: N,
+}
+
+impl<N> RotatingFileLoggerBuilder<N>
+where
+    N: 'static + Send + Fn(&Path) -> io::Result<PathBuf>,
+{
+    pub fn new(path: impl AsRef<Path>, rename: N) -> Self {
+        RotatingFileLoggerBuilder {
+            path: path.as_ref().to_path_buf(),
+            rotators: vec![],
+            rename,
+        }
+    }
+
+    pub fn add_rotator<R: 'static + Rotator>(mut self, rotator: R) -> Self {
+        if rotator.is_enabled() {
+            self.rotators.push(Box::new(rotator));
+        }
+        self
+    }
+
+    pub fn build(mut self) -> io::Result<RotatingFileLogger<N>> {
+        let file = open_log_file(&self.path)?;
+
+        for rotator in self.rotators.iter_mut() {
+            rotator.prepare(&file)?;
+        }
+
+        Ok(RotatingFileLogger {
+            rotators: self.rotators,
+            path: self.path,
+            rename: self.rename,
             file,
         })
     }
+}
 
-    fn should_rotate(&mut self) -> bool {
-        Utc::now() > self.next_rotation_time
+impl<N> Write for RotatingFileLogger<N>
+where
+    N: 'static + Send + Fn(&Path) -> io::Result<PathBuf>,
+{
+    fn write(&mut self, bytes: &[u8]) -> io::Result<usize> {
+        // Updates all roators' states.
+        for rotator in self.rotators.iter_mut() {
+            rotator.on_write(bytes)?;
+        }
+        self.file.write(bytes)
     }
 
-    fn rename(&self) -> io::Result<PathBuf> {
-        fail_point!("file_log_rename", |t| {
-            if let Some(t) = t {
-                Err(match t.as_ref() {
-                    "NotFound" => ErrorKind::NotFound,
-                    "PermissionDenied" => ErrorKind::PermissionDenied,
-                    "AlreadyExists" => ErrorKind::AlreadyExists,
-                    "InvalidInput" => ErrorKind::InvalidInput,
-                    "InvalidData" => ErrorKind::InvalidData,
-                    "WriteZero" => ErrorKind::WriteZero,
-                    "UnexpectedEof" => ErrorKind::UnexpectedEof,
-                    _ => ErrorKind::Other,
+    fn flush(&mut self) -> io::Result<()> {
+        if self.rotators.is_empty() {
+            return self.file.flush();
+        }
+        for rotator in self.rotators.iter() {
+            if rotator.should_rotate()? {
+                let new_path = (self.rename)(&self.path)?;
+                rotator.rotate(&mut self.file, &self.path, &new_path)?;
```

Comment:
Looks like this don't need to be a `rotator.rotate`. Its content can be simply extracted here.

Reply:
Done